### PR TITLE
DCS-809 Redirect when no availability on submitting update

### PR DESCRIPTION
**Pull request by andrewrlee**
Based on the result of the availability check just before submit we now redirect to one of 2 places.

1.) Room no longer available: if someone has booked one or more of the rooms that the user selected

2.) No rooms available: if someone has come along and booked the only remaining room

### CODE REVIEW

File: backend/services/model.ts [hunk 1]
```diff
@@ -17,11 +17,17 @@ export type AvailabilityRequest = {
...
+export type AvailabilityStatus = 'AVAILABLE' | 'NOT_AVAILABLE' | 'NO_LONGER_AVAILABLE'
+
 export type LegacyRoomAvailability = { mainLocations: Room[]; preLocations: Room[]; postLocations: Room[] }
 
+export type SelectedRooms = { pre?: number; main: number; post?: number }
+
+export type Rooms = { main: Room[]; pre: Room[]; post: Room[] }
+
 export type RoomAvailability = {
   isAvailable: boolean
```
Comment (bell-pa) [inline backend/services/model.ts:29]: Looking at this made me wonder if some of the logic associated with rooms and room availability could become methods on these types.  Room Availability becomes 1..3x(Room[] + Interval) with isAvailable being the rule about room combinations perhaps. totalInterval - derived from individual intervals... hasPre, hasPost...  Dunno.

**Review by bell-pa — DISMISSED** [review]
LGTM

File: backend/services/model.ts [hunk 2]
```diff
@@ -17,11 +17,17 @@ export type AvailabilityRequest = {
   postRequired: boolean
 }
 
+export type AvailabilityStatus = 'AVAILABLE' | 'NOT_AVAILABLE' | 'NO_LONGER_AVAILABLE'
```
[the remark this answers is not in the document]
Comment (andrewrlee) [reply backend/services/model.ts:20]: they aren't particularly useful - I think they are only useful if you need to enumerate the values at runtime

The unions provide autocompletion, type safety and exhaustiveness checking in switch statements

File: backend/services/model.ts (hunk 1 above)
Comment (andrewrlee) [reply backend/services/model.ts:29]: meh.. would have to use start using classes. 
Tempted to leave as is for now and maybe play around in a follow up.

File: backend/services/bookingService.ts [hunk 3]
```diff
@@ -84,19 +93,38 @@ export = class BookingService {
...
-        : {}),
-      main: this.toAppointment(update.mainLocation, [update.startTime, update.endTime]),
-      ...(update.postLocation
-        ? { post: this.toAppointment(update.postLocation, postAppointmentTime(update.endTime)) }
-        : {}),
-    })
-    const details = await this.get(context, videoBookingId)
-    await this.notificationService.sendBookingUpdateEmails(context, currentUsername, details)
+  ): Promise<AvailabilityStatus> {
+    const bookingDetails = await this.get(context, videoBookingId)
```
Comment (andrewrlee) [inline backend/services/bookingService.ts:97]: If we passed agency Id through with the cookie then we could avoid a couple of additional calls.

**Review by GurnankCheema — APPROVED** [review]
LGTM

**Pull request merged**